src/mapping/mod.rs

```diff
@@ -0,0 +1,157 @@
+use crate::event::Event;
+
+pub mod parser;
+pub mod query;
+
+use string_cache::DefaultAtom as Atom;
+
+pub type Result<T> = std::result::Result<T, String>;
+
+pub trait Function: Send + core::fmt::Debug {
+    fn apply(&self, target: &mut Event) -> Result<()>;
+}
+
+//------------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub struct Assignment {
+    path: String,
+    function: Box<dyn query::Function>,
```

Comment:
Performance wise I'm not concerned but if it makes the codebase cleaner then I'm down for it, if I have time to play around I'll give this a go.

Reply:
@bruceg I looked into this, but I don't think this can be done easily, without some more substantive changes to the implementation. This is because referencing the enum in the relevant types would create an infinite cycle, which we don't have right now due to boxing trait objects:

```
error[E0072]: recursive type `mapping::MappingFunc` has infinite size
  --> src/mapping/mod.rs:13:1
   |
13 | pub(in crate::mapping) enum MappingFunc {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ recursive type has infinite size
...
16 |     IfStatement,
   |     ----------- recursive without indirection
```

I guess we could box the enum as well, but then we lose a big part of the gains from using `enum_dispatch`.

I'm putting this on ice for now, but I'm happy to revisit if it appears my experimental outcome was flawed, or there's a way to prevent the self-referential issues (or we consider `enum_dispatch` an overall win, even with the required boxing).